src/engraving/libmscore/beam.cpp

```diff
@@ -1450,8 +1488,474 @@ void Beam::add8thSpaceSlant(PointF& dictatorAnchor, int dictator, int pointer, i
     if ((_up && (dictator + 4) % 4 == 3) || (!_up && (dictator + 4) % 4 == 1)) {
         return;
     }
-    dictatorAnchor.setY(dictatorAnchor.y() + (_up ? -0.125 * spatium() : 0.125 * spatium()));
-    _beamDist += 0.0625 * spatium();
+    dictatorAnchor.setY(dictatorAnchor.y() + (_up ? -0.125 * _spatium : 0.125 * _spatium));
+    _beamDist += 0.0625 * _spatium;
+}
+
+bool BeamLayout::computeTremoloUp()
+{
+    if (!_beam || !_beam->cross()) {
+        return _trem->up();
+    }
+    Chord* c1 = _trem->chord1();
+    Chord* c2 = _trem->chord2();
+    int staffMove = 0;
+    for (ChordRest* cr : _beam->elements()) {
+        if (cr->staffMove() != 0) {
+            staffMove = cr->staffMove();
+            break;
+        }
+    }
+    // for inset trems in cross-staff beams, there are two potentialities:
+    // 1) both trem notes are on the same staff
+    if (c1->staffMove() == c2->staffMove()) {
+        // (in which case, up is opposite to the move)
+        if (staffMove < 0) {
+            return c1->staffMove() != staffMove;
+        } else {
+            return c1->staffMove() == staffMove;
+        }
+    } else {
+        // or 2) the trem notes are between staves. what a pain.
+        return _trem->up();
+    }
+}
+
+BeamLayout::BeamLayout(EngravingItem* e)
+{
+    bool isGrace = false;
+    if (!e->isBeam() && !e->isTremolo()) {
+        // right now only beams and trems are supported
```

Comment:
Is it possible to split this class into two independent classes (BeamLayout and TremoloLayout)?

Reply:
I feel like the whole point of this PR is to unify beam layout behavior between two-note tremolos and beams. That is to say, everything in this class applies equally to both and the behaviors of both are exactly identical. I have intentionally minimized the amount of code that is not beam-trem agnostic, so as not to need multiple classes.

Reply:
It should be renamed to BeamAndTremoloLayout at least (to show explicitly what engraving elements this class can handle)

Reply:
`BeamTremoloLayout` was my compromise, hope that's okay!